### PR TITLE
Migrated the data pipeline and training loop to use keras 3 APIs in DEIT Tutorial

### PR DESCRIPTION
**Pull request by kharshith-k**
Please find my PR for migrating data pipeline and residual tensorflow specific APIs to keras 3 in Deit tutorial
[Colab Gist](https://colab.research.google.com/gist/kharshith-k/71554b41d1b12d7599c578b3e0626d36/deit.ipynb)

### CODE REVIEW

File: examples/vision/deit.py [hunk 1]
```diff
@@ -92,60 +87,123 @@
...
+            self.rng.shuffle(self.indices)
+
+    def __getitem__(self, idx):
+        start = idx * self.batch_size
+        end = min((idx + 1) * self.batch_size, len(self.image_paths))
+        batch_indices = self.indices[start:end]
+        images = []
+        for i in batch_indices:
+            image = keras.utils.load_img(
+                self.image_paths[i], target_size=(RESOLUTION + 20, RESOLUTION + 20)
```
Comment (sachinprasadhs) [inline examples/vision/deit.py]: You need to address this Gemini comment. 

This will be impact on memory, especially for validation data. Usualy validation data will be shuffle=False and validation does not perform random crop, so this reolution+20 will be inefficient and extra load on disk space.
May be we can add condition like below.

```
target_size = (
    (RESOLUTION + 20, RESOLUTION + 20)
    if self.shuffle
    else (RESOLUTION, RESOLUTION)
)
image = keras.utils.load_img(
    self.image_paths[i], target_size=target_size
)
```

File: examples/vision/deit.py [hunk 2]
```diff
@@ -482,124 +454,81 @@ def metrics(self):
...
+    [teacher_backbone, layers.Dense(NUM_CLASSES)], name="teacher"
+)
+teacher_model.compile(
+    optimizer=keras.optimizers.AdamW(learning_rate=1e-3, weight_decay=1e-4),
+    loss=keras.losses.CategoricalCrossentropy(from_logits=True),
+    metrics=[keras.metrics.CategoricalAccuracy(name="accuracy")],
+)
+
+print("Fine-tuning teacher head on flowers dataset...")
+teacher_model.fit(train_dataset, validation_data=val_dataset, epochs=5)
```
Comment (sachinprasadhs) [inline examples/vision/deit.py:531]: Let's add workers and set some value and if required set use_multiprocessing args
Same thing for `deit_distiller.fit`

**Review by sachinprasadhs — COMMENTED** [review]
Added some performance related comments, please address them and regenerate files. Thanks

**Review by sachinprasadhs — APPROVED** [review]
